File: compiler-core/src/main/java/com/siimkinks/sqlitemagic/processing/DatabaseConfigurationCollectionStep.java
package com.siimkinks.sqlitemagic.processing;

import com.google.common.base.Strings;
import com.siimkinks.sqlitemagic.BaseProcessor;
import com.siimkinks.sqlitemagic.Environment;
import com.siimkinks.sqlitemagic.annotation.Database;
import com.siimkinks.sqlitemagic.annotation.SubmoduleDatabase;
import com.siimkinks.sqlitemagic.util.Dual;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import javax.annotation.processing.RoundEnvironment;
import javax.inject.Inject;
import javax.lang.model.element.Element;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.MirroredTypesException;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;

import static com.siimkinks.sqlitemagic.Environment.getGenClassesManagerClassName;
import static com.siimkinks.sqlitemagic.util.NameConst.PACKAGE_ROOT;
import static com.siimkinks.sqlitemagic.util.StringUtil.firstCharToUpperCase;

public class DatabaseConfigurationCollectionStep implements ProcessingStep {
  private static final String ERR_SINGLE_SUBMODULE_DB_ALLOWED = String
      .format("Only one element per module can be annotated with @%s", SubmoduleDatabase.class.getSimpleName());
  private static final String ERR_SINGLE_DB_ALLOWED = String
      .format("Only one element per module can be annotated with @%s", Database.class.getSimpleName());

  @Inject
  Environment environment;

  public DatabaseConfigurationCollectionStep() {
    BaseProcessor.inject(this);
  }

  @Override
  public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
    final Set<? extends Element> submoduleDatabaseElements = roundEnv.getElementsAnnotatedWith(SubmoduleDatabase.class);
    if (!parseSubmodules(submoduleDatabaseElements)) return false;

    final Set<? extends Element> databaseElements = roundEnv.getElementsAnnotatedWith(Database.class);
    if (!submoduleDatabaseElements.isEmpty() && !databaseElements.isEmpty()) {
      environment.error("Module can have either @%s or @%s annotated element, but not both",
          Database.class.getSimpleName(), SubmoduleDatabase.class.getSimpleName());
      return false;
    }
    if (databaseElements.size() > 1) {
      environment.error(ERR_SINGLE_DB_ALLOWED);
      return false;
    }
    for (Element element : databaseElements) {
      final Database database = element.getAnnotation(Database.class);
      if (environment.getSubmoduleDatabases() == null) {
        final List<Dual<TypeElement, String>> submodules = getSubmodules(environment, database);
        if (!submodules.isEmpty()) {
          environment.setSubmoduleDatabases(submodules);
        }
      } else {
        environment.error(element, ERR_SINGLE_DB_ALLOWED);
        return false;
      }
      final String dbName = database.name();
      if (!Strings.isNullOrEmpty(dbName)) {
        environment.setDbName("\"" + dbName + "\"");
      }
      final int version = database.version();
      if (version >= 0) {
        environment.setDbVersion(Integer.toString(version));
      }
    }

    return true;
  }

  private boolean parseSubmodules(Set<? extends Element> submoduleDatabaseElements) {
    if (submoduleDatabaseElements.size() > 1) {
      environment.error(ERR_SINGLE_SUBMODULE_DB_ALLOWED);
      return false;
    }
    for (Element element : submoduleDatabaseElements) {
      if (environment.getModuleName() == null) {
        final SubmoduleDatabase submoduleDatabase = element.getAnnotation(SubmoduleDatabase.class);
        final String moduleName = submoduleDatabase.value();
        if (Strings.isNullOrEmpty(moduleName)) {
          environment.error(element, "Submodule name cannot be empty or null");
        }
        environment.setModuleName(firstCharToUpperCase(moduleName));
      } else {
        environment.error(element, ERR_SINGLE_SUBMODULE_DB_ALLOWED);
        return false;
      }
    }
    return true;
  }

  private static List<Dual<TypeElement, String>> getSubmodules(Environment environment, Database databaseAnnotation) {
    final ArrayList<Dual<TypeElement, String>> result = new ArrayList<>();
    final Elements elementUtils = environment.getElementUtils();
    try {
      for (Class<?> clazz : databaseAnnotation.submodules()) {
        final TypeElement annotatedTypeElement = environment.getTypeElement(clazz);
        final Dual<TypeElement, String> managerElement = getManagerTypeElement(elementUtils, annotatedTypeElement);
        result.add(managerElement);
      }
    } catch (MirroredTypesException e) {
      for (TypeMirror typeMirror : e.getTypeMirrors()) {
        final TypeElement annotatedTypeElement = elementUtils.getTypeElement(typeMirror.toString());
        final Dual<TypeElement, String> managerElement = getManagerTypeElement(elementUtils, annotatedTypeElement);
        result.add(managerElement);
      }
    }
    return result;
  }

  private static Dual<TypeElement, String> getManagerTypeElement(Elements elementUtils, TypeElement annotatedTypeElement) {
    final SubmoduleDatabase submoduleDatabase = annotatedTypeElement.getAnnotation(SubmoduleDatabase.class);
    final String moduleName = firstCharToUpperCase(submoduleDatabase.value());
    final String managerClassName = getGenClassesManagerClassName(moduleName);
    return Dual.create(
        elementUtils.getTypeElement(PACKAGE_ROOT + "." + managerClassName),
        moduleName);
  }
}

File: compiler-ksp/src/main/kotlin/com/siimkinks/sqlitemagic/dbconfig/DatabaseConfigurationCollectionStep.kt
package com.siimkinks.sqlitemagic.dbconfig

import com.google.devtools.ksp.KSTypesNotPresentException
import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSAnnotated
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.siimkinks.sqlitemagic.Environment
import com.siimkinks.sqlitemagic.NameConst
import com.siimkinks.sqlitemagic.Types
import com.siimkinks.sqlitemagic.annotation.Database
import com.siimkinks.sqlitemagic.annotation.SubmoduleDatabase
import com.siimkinks.sqlitemagic.processing.ProcessingStep
import com.siimkinks.sqlitemagic.processing.ProcessingStepResult
import com.siimkinks.sqlitemagic.utils.findAnnotationWithType
import com.siimkinks.sqlitemagic.utils.firstCharToUpperCase

class DatabaseConfigurationCollectionStep(
  private val environment: Environment
) : ProcessingStep {
  override fun process(resolver: Resolver): ProcessingStepResult {
    val submoduleDatabaseSymbols = resolver
      .getSymbolsWithAnnotation(Types.SUBMODULE_DATABASE_ANNOTATION)
      .toList()
    if (!parseSubmodules(submoduleDatabaseSymbols)) {
      return ProcessingStepResult.Failed
    }

    val databaseSymbols = resolver
      .getSymbolsWithAnnotation(Types.DATABASE_ANNOTATION)
      .toList()
    if (submoduleDatabaseSymbols.isNotEmpty() && databaseSymbols.isNotEmpty()) {
      environment.logger.error(
        message = "Module can have either $DATABASE_ANNOTATION_NAME or $SUBMODULE_DATABASE_ANNOTATION_NAME annotated element, but not both"
      )
      return ProcessingStepResult.Failed
    }
    if (databaseSymbols.size > 1) {
      environment.logger.error(ERR_SINGLE_DB_ALLOWED)
      return ProcessingStepResult.Failed
    }
    for (symbol in databaseSymbols) {
      if (environment.submoduleDatabases != null) {
        environment.logger.error(
          message = ERR_SINGLE_DB_ALLOWED,
          symbol = symbol
        )
        return ProcessingStepResult.Failed
      }
      val databaseAnnotation = checkNotNull(symbol.findAnnotationWithType<Database>())
      val submodules = resolver
        .getSubmodules(
          databaseAnnotation = databaseAnnotation,
          databaseSymbol = symbol
        )
        ?: return ProcessingStepResult.Failed
      if (submodules.isNotEmpty()) {
        environment.submoduleDatabases = submodules
      }
      environment.setDatabaseMetadata(
        dbName = databaseAnnotation.name,
        dbVersion = databaseAnnotation.version
      )
    }
    return ProcessingStepResult.Continue
  }

  private fun parseSubmodules(submoduleDatabaseSymbols: List<KSAnnotated>): Boolean {
    if (submoduleDatabaseSymbols.size > 1) {
      environment.logger.error(ERR_SINGLE_SUBMODULE_DB_ALLOWED)
      return false
    }
    for (symbol in submoduleDatabaseSymbols) {
      if (environment.submoduleName != null) {
        environment.logger.error(
          message = ERR_SINGLE_SUBMODULE_DB_ALLOWED,
          symbol = symbol
        )
        return false
      }

      val submoduleDatabase = checkNotNull(symbol.findAnnotationWithType<SubmoduleDatabase>())
      val moduleName = submoduleDatabase.value
      if (moduleName.isEmpty()) {
        environment.logger.error(
          message = "Submodule name cannot be empty or null",
          symbol = symbol
        )
        return false
      }
      environment.setSubmoduleName(moduleName)
    }
    return true
  }

  private fun Resolver.getSubmodules(
    databaseAnnotation: Database,
    databaseSymbol: KSAnnotated
  ): List<SubmoduleDatabaseMetadata>? = buildList {
    val submoduleDeclarations = getSubmoduleDeclarations(
      databaseAnnotation = databaseAnnotation,
      databaseSymbol = databaseSymbol
    ) ?: return null
    for (submoduleDeclaration in submoduleDeclarations) {
      add(getSubmoduleMetadata(submoduleDeclaration) ?: return null)
    }
  }

  private fun Resolver.getSubmoduleDeclarations(
    databaseAnnotation: Database,
    databaseSymbol: KSAnnotated
  ): List<KSClassDeclaration>? = try {
    databaseAnnotation.submodules.map { submoduleClass ->
      val submoduleQualifiedName = submoduleClass.qualifiedName
        ?: run {
          environment.logger.error(
            message = "Database submodule ${submoduleClass.simpleName} could not be resolved",
            symbol = databaseSymbol
          )
          return null
        }
      getClassDeclarationByName(submoduleQualifiedName)
        ?: run {
          environment.logger.error(
            message = "Database submodule $submoduleQualifiedName could not be resolved",
            symbol = databaseSymbol
          )
          return null
        }
    }
  } catch (e: KSTypesNotPresentException) {
    e.ksTypes.map { submoduleType ->
      submoduleType.declaration as? KSClassDeclaration
        ?: run {
          environment.logger.error(
            message = "Database submodule must be a class: ${submoduleType.declaration.qualifiedName?.asString()}"
          )
          return null
        }
    }
  }

  private fun getSubmoduleMetadata(
    annotatedDeclaration: KSClassDeclaration
  ): SubmoduleDatabaseMetadata? {
    val submoduleDatabase = annotatedDeclaration.findAnnotationWithType<SubmoduleDatabase>()
    if (submoduleDatabase == null) {
      environment.logger.error(
        message = "Database submodule ${annotatedDeclaration.qualifiedName?.asString()} must be annotated with $SUBMODULE_DATABASE_ANNOTATION_NAME",
        symbol = annotatedDeclaration
      )
      return null
    }

    val moduleName = submoduleDatabase.value
    if (moduleName.isEmpty()) {
      environment.logger.error(
        message = "Submodule name cannot be empty",
        symbol = annotatedDeclaration
      )
      return null
    }

    val capitalizedModuleName = moduleName.firstCharToUpperCase()
    val managerClassName = environment.getGenClassesManagerClassName(capitalizedModuleName)
    return SubmoduleDatabaseMetadata(
      managerQualifiedName = "${NameConst.PACKAGE_ROOT}.$managerClassName",
      moduleName = capitalizedModuleName
    )
  }

  private companion object {
    val DATABASE_ANNOTATION_NAME = "@${Database::class.simpleName}"
    val SUBMODULE_DATABASE_ANNOTATION_NAME = "@${SubmoduleDatabase::class.simpleName}"
    val ERR_SINGLE_SUBMODULE_DB_ALLOWED =
      "Only one element per module can be annotated with $SUBMODULE_DATABASE_ANNOTATION_NAME"
    val ERR_SINGLE_DB_ALLOWED = "Only one element per module can be annotated with $DATABASE_ANNOTATION_NAME"
  }
}